### PR TITLE
dynamic bins samples for speeding up

### PR DESCRIPTION
**Pull request by namisan**

### CODE REVIEW

File: mt_dnn/batcher.py [hunk 1]
```diff
@@ -34,6 +53,27 @@ def _get_shuffled_index_batches(dataset_len, batch_size):
...
+        maxlen = dataset._maxlen
+        bins = create_bins(bin_size, maxlen)
+        data = [[] for i in range(0, len(bins))]
+        
+        for idx, sample in enumerate(dataset):
+            bin_idx = search_bin(bins, len(sample['sample']['token_id']))
+            data[bin_idx].append(idx)
+        index_batches = []
+
+        for idx, sub_data in enumerate(data):
```
Comment (anselmwang) [inline mt_dnn/batcher.py:67]: the length of bins are [64, 128, 192, 256, 320, 384, 448, 512], which increases linearly,
but the batch_size scales down exponentially by "batch_size * bin_grow_ratio", is this the optimal?
I guess the memory grows linearly with O(len) , then could batch_size = batch_size_for_64 // cur_bin_seq_len?

File: mt_dnn/batcher.py [hunk 2]
```diff
@@ -34,6 +53,27 @@ def _get_shuffled_index_batches(dataset_len, batch_size):
         random.shuffle(index_batches)
         return index_batches
 
+    @staticmethod
+    def _get_shuffled_index_batches_bin(dataset, batch_size, bin_size, bin_grow_ratio):
+        maxlen = dataset._maxlen
```
Comment (anselmwang) [inline mt_dnn/batcher.py:58]: how about we expose _maxlen as public (rename to be "maxlen") ? I didn't aware _maxlen would be access by other class

File: mt_dnn/batcher.py [hunk 3]
```diff
@@ -34,6 +53,27 @@ def _get_shuffled_index_batches(dataset_len, batch_size):
...
+        data = [[] for i in range(0, len(bins))]
+        
+        for idx, sample in enumerate(dataset):
+            bin_idx = search_bin(bins, len(sample['sample']['token_id']))
+            data[bin_idx].append(idx)
+        index_batches = []
+
+        for idx, sub_data in enumerate(data):
+            if len(sub_data) < 1: continue
+            batch_size = int(batch_size * bin_grow_ratio)
```
Comment (anselmwang) [inline mt_dnn/batcher.py:69]: from current code, the batch size of shortest seq is int(batch_size * bin_grow_ratio) instead of batch_size, this has 2 disadvantages:
1. confusing.
2. even the shortest seq batch (largest batch) 's loss is scaled down with code in model.py "loss = loss * (1.0 * batch_size / self.config['batch_size'])"

better to move line 69-70 after line 73

File: mt_dnn/batcher.py (hunk 1 above)
Comment (anselmwang) [reply mt_dnn/batcher.py:67]: if O(len^2) batch_size = batch_size_for_64 // (128*128// (64*64))

File: mt_dnn/batcher.py (hunk 3 above)
Comment (namisan) [reply mt_dnn/batcher.py]: thanks, fixed.
